File: 2. Linked List/Double Linked List/DoublyLinkedList.java
import java.util.Scanner;

public class DoublyLinkedList {
    private int data;
    private DoublyLinkedList next, prev;

    private int listLenght = 0;
    private DoublyLinkedList head = null, tail = null;

    private DoublyLinkedList createLinkedListNode(int val) {
        try {
            DoublyLinkedList node = new DoublyLinkedList();
            node.data = val;
            node.next = null;
            node.prev = null;
            return node;
        } catch (Exception e) {
            System.out.println("Failed Allocation");
        }
        return null;
    }

    private void insert(int val, int pos) {
        if (pos <= 0 | pos > listLenght + 1) {
            System.out.println("~~ Invalid Position ~~");
        }

        DoublyLinkedList newNode = createLinkedListNode(val);
        // 4 -> 6 -> 9 -> n -> 12 -> 3 -> 3 -> NULL / 7
        // NUll <- 4 <- 6 <- 9 <- n <- 12 <- 3 <- 3 / 7
        if (pos == 1) {
            if (head != null) {
                newNode.next = head;
                head.prev = newNode;
            }

            newNode.prev = null;
            head = newNode;

            if (tail == null) {
                tail = head;
            } else if (tail.prev == null) {
                tail.prev = head;
            }
        } else if (pos == listLenght + 1) {
            newNode.next = null;
            newNode.prev = tail;
            tail.next = newNode;
            tail = newNode;
        }  else {
            DoublyLinkedList dummy = head;
            int i = 1;
            while (i < pos - 1) {
                i++;
                dummy = dummy.next;
            }
            newNode.next = dummy.next;
            newNode.prev = dummy;
            dummy.next.prev = newNode;
            dummy.next = newNode;
        }
        listLenght++;
        System.out.printf("Insertion Successful at Position:- %d. Value:- %d\n\n", pos, val);
    }

    public void delete(int pos) {
        if (pos <= 0 | pos > listLenght) {
            System.out.println("~~ Invalid Position ~~");
        }

        int toBeDeletedValue;
        DoublyLinkedList toBeDeletedNode;
        // 4 -> 6 -> 9 -> n -> 12 -> 3 -> 3 -> NULL / 7
        // NUll <- 4 <- 6 <- 9 <- n <- 12 <- 3 <- 3 / 7
        if (pos == 1) {
            toBeDeletedNode = head;
            toBeDeletedValue = toBeDeletedNode.data;
            head = toBeDeletedNode.next;
            head.prev = null;
        } else if (pos == listLenght) {
            toBeDeletedNode = tail;
            toBeDeletedValue = toBeDeletedNode.data;
            tail = toBeDeletedNode.prev;
            tail.next = null;
        }  else {
            DoublyLinkedList dummy = head;
            int i = 1;
            while (i < pos - 1) {
                i++;
                dummy = dummy.next;
            }
            toBeDeletedNode = dummy.next;
            toBeDeletedValue = toBeDeletedNode.data;
            dummy.next = toBeDeletedNode.next;
            toBeDeletedNode.next.prev = toBeDeletedNode.prev;
        }

        listLenght--;
        System.out.printf("\nDeletion Successful at Position:- %d. Value:- %d\n\n", pos, toBeDeletedValue);
    }

    public void print(DoublyLinkedList list ,boolean reverse){
        if (reverse) {
            String listStr = Integer.toString(list.tail.data);
            DoublyLinkedList dum = list.tail;
            while (dum.prev != null) {
                dum = dum.prev;
                listStr = listStr + " -> " + Integer.toString(dum.data);
            }
            System.out.printf("\nLinked List(Reverse):- %s -> NULL\n\n", listStr);
        } else {
            String listStr = Integer.toString(list.head.data);
            DoublyLinkedList dum = list.head;
            while (dum.next != null) {
                dum = dum.next;
                listStr = listStr + " -> " + Integer.toString(dum.data);
            }
            System.out.printf("\nLinked List:- %s -> NULL\n\n", listStr);
        }
    }

    public static void main(String[] args) {
        Scanner myObj = new Scanner(System.in);
        DoublyLinkedList list = new DoublyLinkedList();

        int choice, val, pos;
        System.out.println("Hello! Here You Can Perform Following Linked List Operations!");
        while (true) {
            System.out.println("1. Show Data\n2. Show Data(Reverse)\n3. Insert at First\n4. Insert At Last");
            System.out.println("5. Insert at Specific Position\n6. Remove From First\n7. Remove From Last");
            System.out.println("8. Remove From Specific Position\n9. Length of Data\n10. Exit");
            System.out.print("Enter Your Choice: ");
            choice = myObj.nextInt();

            switch (choice) {
                case 1:
                    if (list.listLenght == 0) {
                        System.out.println("\n~~Linked List is Empty~~\n");
                        break;
                    }
                    list.print(list, false);
                    break;
                case 2:
                    if (list.listLenght == 0) {
                        System.out.println("\n~~Linked List is Empty~~\n");
                        break;
                    }
                    list.print(list, true);
                    break;
                case 3:
                    System.out.print("\nEnter Number: ");
                    val = myObj.nextInt();
                    list.insert(val, 1);
                    break;
                case 4:
                    System.out.print("\nEnter Number: ");
                    val = myObj.nextInt();
                    list.insert(val, list.listLenght + 1);
                    break;
                case 5:
                    System.out.print("\nEnter Number: ");
                    val = myObj.nextInt();
                    System.out.print("Enter Position(Greater than 0): ");
                    pos = myObj.nextInt();
                    list.insert(val, pos);
                    break;
                case 6:
                    if (list.listLenght == 0) {
                        System.out.println("\n~~Linked List is Empty~~\n");
                        break;
                    }
                    list.delete(1);
                    break;
                case 7:
                    if (list.listLenght == 0) {
                        System.out.println("\n~~Linked List is Empty~~\n");
                        break;
                    }
                    list.delete(list.listLenght);
                    break;
                case 8:
                    if (list.listLenght == 0) {
                        System.out.println("\n~~Linked List is Empty~~\n");
                        break;
                    }
                    System.out.print("\nEnter Position(Greater than 0): ");
                    pos = myObj.nextInt();
                    list.delete(pos);
                    break;
                case 9:
                    System.out.printf("\nLinked List Length: %d\n\n", list.listLenght);
                    break;
                case 10:
                    System.out.println("\n~~Thanks for Using! Have a Great Day!~~");
                    myObj.close();
                    return;
                default:
                    System.out.println("\nPlease Enter a Valid Input!");
            }
        }
    }
}

File: 2. Linked List/Double Linked List/doublyLinkedList.py
head = None;
tail = None;
listLength = 0;

class linkedList:
    def __init__(self, data, next, prev):
        self.data = data;
        self.next = next;
        self.prev = prev;

def createLinkedListNode(val, next, prev):
    node = linkedList(val, next, prev);
    global listLength;
    listLength += 1;
    return node;

def insert(val, pos):
    global listLength, head, tail;
    if pos <= 0 | pos > listLength + 1:
        print("\n~~Invalid Position: ", pos, "~~\n");
        return;

    newNode = createLinkedListNode(val, None, None);

    if (pos == 1):
        newNode.next = head;
        head = newNode;
        if (tail == None):
            tail = head;
        elif (tail.prev == None):
            tail.prev = head;
    elif (pos == listLength):
        tail.next = newNode;
        newNode.prev = tail;
        tail = newNode;
        if (head == None):
            head.next = newNode;
    else:
        prev = head;
        i = 1;
        while i < pos - 1:
            i += 1
            prev = prev.next;
        newNode.next = prev.next;
        newNode.prev = prev;
        prev.next.prev = newNode;
        prev.next = newNode;
    print("Insertion Successful at Position:- ", pos, ". Value:- ", val, "\n");

def delete(pos):
    global listLength, head, tail;
    if pos <= 0 | pos > listLength:
        print("\n~~Invalid Position: ", pos, "~~\n");
        return;

    toBeRemovedValue = None;
# 1 -> 7 -> 4 -> 9 -> 0 -> 1 -> N 
    if (pos == 1):
        toBeRemovedValue = head.data;
        head = head.next;
        head.prev = None;
    elif (pos == listLength):
        toBeRemovedValue = tail.data;
        tail = tail.prev;
        tail.next = None;
    else:
        prev = head;
        i = 1;
        while i < pos - 1:
            i += 1
            prev = prev.next;
        toBeRemovedValue = prev.next.data;
        prev.next.next.prev = prev;
        prev.next = prev.next.next;
    listLength -= 1;
    print("Deletion Successful at Position:- ", pos, ". Value:- ", toBeRemovedValue, "\n")

print("Hello! Here You Can Perform Following Linked List Operations!");
while (True) :
    print("1. Show Data\n2. Show Data(Reverse)\n3. Insert at First\n4. Insert at Last\n5. Insert at Specific Position\n6. Remove from First\n7. Remove from Last\n8. Remove from Specific Position\n9. Length of Data\n10. Exit\n");
    choice = int(input("Enter Your Choice:- "));
    if(choice == 1):
        if (head == None):
            print("\n~~ Linked List is Empty ~~\n");
            continue;
        listString = str(head.data);
        dum = head;
        while(dum.next != None):
            dum = dum.next
            listString = listString + ' -> ' + str(dum.data);
        print("\nLinked List:", listString, "-> NULL\n");
        continue;
    elif(choice == 2):
        if (tail == None):
            print("\n~~ Linked List is Empty ~~\n");
            continue;
        listString = str(tail.data);
        dum = tail;
        while(dum.prev != None):
            dum = dum.prev
            listString = listString + ' -> ' + str(dum.data);
        print("\nLinked List(Reverse):", listString, "-> NULL\n");
        continue;
    elif(choice == 3):
        num = int(input("\nEnter Your Number:- "));
        insert(num, 1);
        continue;
    elif(choice == 4):
        num = int(input("\nEnter Your Number:- "));
        insert(num, listLength + 1);
        continue;
    elif(choice == 5):
        num = int(input("\nEnter Your Number:- "));
        pos = int(input("Enter Your Position(Greater Than 0):- "));
        insert(num, pos);
        continue;
    elif(choice == 6):
        print('\n');
        delete(1);
        continue;
    elif(choice == 7):
        print('\n');
        delete(listLength);
        continue;
    elif(choice == 8):
        pos = int(input("\nEnter Your Position(Greater Than 0):- "));
        delete(pos);
        continue;
    elif(choice == 9):
        print("\nList Lenght:- ", listLength, "\n");
        continue;
    elif(choice == 10):
        print("\n~~ Thanks For Using! Have a Great Day! ~~\n");
        break;
    else:
        print("\n~~ Please Enter a Vaild Input! ~~\n");
        continue;
    
